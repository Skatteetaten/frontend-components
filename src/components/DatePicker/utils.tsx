export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
