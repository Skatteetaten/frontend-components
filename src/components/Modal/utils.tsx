import { useEffect } from 'react';

export const createModalDomPlacement = (
  setFlag: (a: boolean) => void,
  shadowRootNode: Document | ShadowRoot | undefined
) => {
  if (!shadowRootNode) {
    setFlag(true);
    return;
  }

  const styleInjectorWrapperId = 'style-injector';
  const modalWrapperId = 'modal-wrapper';

  if (shadowRootNode.getElementById(modalWrapperId)) {
    setFlag(true);
  } else {
    const modalWrapper = document.createElement('div');
    modalWrapper.id = modalWrapperId;
    const styleInjectorWrapper = shadowRootNode.getElementById(
      styleInjectorWrapperId
    );
    const parent = styleInjectorWrapper ? styleInjectorWrapper : shadowRootNode;
    parent.appendChild(modalWrapper);
    setFlag(true);
  }
};

export const useEscOnPress = (
  ref: HTMLDivElement | null,
  onClose?: (ref: HTMLDivElement) => void
): void => {
  useEffect(() => {
    const onKeyDown = (event: KeyboardEvent): void => {
      if (ref == null) {
        return;
      }
      if (event.code === 'Escape' && onClose) {
        onClose(ref);
      }
    };

    if (ref) {
      window.addEventListener('keydown', onKeyDown as EventListener, false);
    }
    return function cleanup(): void {
      if (window) {
        window.removeEventListener('keydown', onKeyDown as EventListener);
      }
    };
  }, [ref, onClose]);
};
