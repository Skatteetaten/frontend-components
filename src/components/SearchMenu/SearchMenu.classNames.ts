import { SearchMenuProps } from './SearchMenu';
import { mergeStyleSets } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';
import { PaletteProps } from '..';

export const getClassNames = (props: SearchMenuProps) => {
  const palette = getTheme().palette as PaletteProps;
  return mergeStyleSets({
    searchBoxWrapper: {
      backgroundColor: palette.skeColor.whiteGrey,
      padding: '12px',
      border: `1px solid ${palette.skeColor.black}`,
      selectors: {
        '.ms-SearchBox': {
          border: 'inherit'
        }
      }
    },
    searchMenuDropdown: {
      marginTop: '-16px',
      paddingTop: '16px',
      selectors: {
        ul: {
          listStyleType: 'none',
          padding: 0
        },
        li: {
          paddingLeft: '30px',
          cursor: 'pointer',
          color: `${palette.skeColor.blackAlt} !important`,
          selectors: {
            ':hover': {
              background: palette.skeColor.lightBlue,
              textDecoration: 'underline'
            },
            ':active': {
              background: palette.skeColor.lightBlue,
              textDecoration: 'none'
            },
            ':focus': {
              background: palette.skeColor.lightBlue,
              textDecoration: 'underline'
            }
          }
        }
      }
    },
    typeCheckbox: {
      cursor: 'default !important'
    }
  });
};
