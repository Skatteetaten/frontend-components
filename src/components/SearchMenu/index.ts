export { default } from './SearchMenu';
export * from './SearchMenu';
