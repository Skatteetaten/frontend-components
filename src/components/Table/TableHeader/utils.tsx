import { TFunction } from 'react-i18next';

export const getIconNameForTheadCell = (
  isSorted: boolean,
  isSortedAscending: boolean
) => {
  if (isSorted) {
    return isSortedAscending ? 'ArrowDown' : 'ArrowUp';
  }

  return 'ArrowUpDown';
};

export const getAriaLabelForTheadCell = (
  name: string | JSX.Element,
  fieldName: string,
  isSorted: boolean,
  isSortedAscending: boolean,
  t: TFunction<Array<string>>
) => {
  let ariaLabel = fieldName;
  if (typeof name === 'string') {
    ariaLabel = isSorted
      ? isSortedAscending
        ? name.concat(' ', t('table.sorted_ascending'))
        : name.concat(' ', t('table.sorted_descending'))
      : name.concat(' ', t('table.sortable'));
  }

  return ariaLabel;
};
