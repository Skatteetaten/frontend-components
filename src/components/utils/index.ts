export { DefaultPalette as Palette, skeColor, skePalette } from './palette';
export type { PaletteProps, SkeColorProp } from './palette';
export {
  Fonts,
  FontSizes,
  LineHeightSizes,
  FontWeights,
  IconFontSizes,
  createFont,
} from './fonts';
export { omstrukturerFlatJSON, t } from './i18n/i18n';
export { getSrOnlyStyle } from './getSrOnlyStyle';
export { languages } from './i18n/languages';
export type { FontSizeType } from './fonts';
export { getFocusStyle } from './getFocusStyle';
export { Heading } from './Heading';
export { UseScreen } from './ScreenPlugin';
export { default as SkeIcons } from './icons/icons-ske';
export { Animation } from './getAnimationStyles';
export { takeIf } from './helpers';
export { generateId } from './generateId';
export { useHotkeys } from './useHotkeys';
