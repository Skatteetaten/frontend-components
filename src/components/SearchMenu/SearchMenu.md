** SearchMenu er en komponent som filtrer menyvalg **

SearchMenu er basert på at de søkbare elementene er pakket inn i en ul og at de har en [ref](https://reactjs.org/docs/refs-and-the-dom.html)
```js
import React from 'react';
import ActionButton from '@skatteetaten/frontend-components/ActionButton';
import SearchMenu from '@skatteetaten/frontend-components/SearchMenu';
import Typography from '@skatteetaten/frontend-components/Typography';


const list_1 = [
  {key: 1, text: 'Forskuddsskatt'},
  {key: 2, text: 'Skattemelding'},
  {key: 3, text: 'Skatteoppgjør'}
];

const list_2 = [
  {key: 4, text: 'Tvangsmulkt'},
  {key: 5, text: 'Tilleggsskatt'},
  {key: 6, text: 'Overtredelsesgebyr'}
];
const [value, setValue] = React.useState(undefined);
const [visible, setVisible] = React.useState(true);
let ref = [];

<div style={{ width: '350px' }}>
    <SearchMenu
      value={value}
      dropdownVisible={visible}
    >
      <Typography>
        <h4 style={{ marginLeft: '20px' }}>Skatt for bedrifter og org</h4>
      </Typography>
      <ul>
        {list_1.map(listItem => {
          ref[listItem.text] = React.useRef();
          return (
            <li key={listItem.key} ref={ref[listItem.text]}>
              <ActionButton
                ariaLabel={listItem.text}
                color={'black'}
                onClick={() => {
                  setValue(listItem.text);
                  setVisible(false);
                }}
              >
                {listItem.text}
              </ActionButton>
            </li>
          );
        })}
      </ul>
      <Typography>
        <h4 style={{ marginLeft: '20px' }}>
          Frister, gebyrer og tilleggsskatt
        </h4>
      </Typography>
      <ul>
        {list_2.map(listItem => {
          ref[listItem.text] = React.useRef();
          return (
            <li key={listItem.key} ref={ref[listItem.text]}>
              <ActionButton
                ariaLabel={listItem.text}
                color={'black'}
                onClick={() => {
                  setValue(listItem.text);
                  setVisible(false);
                }}
              >
                {listItem.text}
              </ActionButton>
            </li>
          );
        })}
      </ul>
    </SearchMenu>
</div>;
```