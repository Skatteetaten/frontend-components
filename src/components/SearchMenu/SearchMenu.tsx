import * as React from 'react';
import {
  ISearchBoxProps,
  SearchBox
} from 'office-ui-fabric-react/lib-commonjs/SearchBox';
import classnames from 'classnames';
import { getClassNames } from './SearchMenu.classNames';
import { getClassNames as searchFieldClasses } from '../SearchField/SearchField.classNames';
import { ReactElement } from 'react';

export interface SearchMenuProps extends ISearchBoxProps {
  /** Boolean for om dropdownlisten skal vises initielt */
  dropdownVisible?: boolean;
  children: Array<ReactElement> | ReactElement;
  /** Boolean for on om listeelementene er av type checkbox */
  checkbox?: boolean;
}

const SearchMenu: React.FC<SearchMenuProps> = props => {
  const {
    checkbox,
    children,
    className,
    dropdownVisible: dropdownVisibleFromProp,
    onChange,
    value: valueFromProp,
    ...rest
  } = props;
  const styles = getClassNames(props);
  const [value, setValue] = React.useState<string | undefined>(valueFromProp);
  const [dropdownVisible, setDropdownVisible] = React.useState<boolean>(
    dropdownVisibleFromProp !== undefined ? dropdownVisibleFromProp : true
  );
  const [focus, setFocus] = React.useState<number>(-1);
  const listRefs = React.useRef<(HTMLLIElement | null)[]>([]);

  const setSearchResult = (newValue: string) => {
    if (children && newValue) {
      const newList = searchInList(newValue);
      setDropdownVisible(newList.length > 0);
    }
  };

  const changeEvent = (text: string | undefined) => {
    //@ts-ignore TODO
    const event: React.ChangeEvent<HTMLInputElement> = {};
    setValue(text);
    onChange && onChange(event, text);
    setDropdownVisible(false);
    setFocus(-1);
    listRefs.current = [];
  };

  const findListItems = (
    childrenInList: Array<ReactElement> | ReactElement
  ) => {
    if (!Array.isArray(childrenInList)) {
      childrenInList = [childrenInList];
    }
    const filtered = childrenInList.map((child: ReactElement, key: number) => {
      if (child.type === 'ul') {
        const grandchildrenList: Array<ReactElement> = [];
        child.props.children.map(
          (grandchildren: ReactElement, childKey: number) => {
            const grandChildrenKey = grandchildren.key
              ? (grandchildren.key as number) - 1
              : childKey;
            const editedGrandchild = (
              <li
                key={grandChildrenKey}
                ref={(ref: HTMLLIElement | null) => {
                  listRefs.current.push(ref);
                }}
                className={classnames(checkbox && styles.typeCheckbox)}
                onClick={() => {
                  const currentKey = listRefs.current[grandChildrenKey];
                  if (!checkbox && currentKey) {
                    changeEvent(currentKey.innerText);
                  }
                }}
                onKeyPress={ev => {
                  const currentKey = listRefs.current[grandChildrenKey];
                  if (ev.keyCode === 0 && currentKey) {
                    changeEvent(currentKey.innerText);
                  }
                }}
                onKeyDown={ev => handleOnKeyDown(ev)}
                tabIndex={0}
              >
                {grandchildren.props.children}
              </li>
            );
            grandchildrenList.push(editedGrandchild);
            return grandchildrenList;
          }
        );
        const clone = React.cloneElement(child, {
          children: grandchildrenList,
          key: child.type.toString().concat(key.toString())
        });
        return clone;
      }
      return child;
    });
    return filtered;
  };

  const searchInList = (filterText: string) => {
    return listRefs.current.filter(item => {
      if (item) {
        if (
          item.innerText
            .replace(/\s/g, '')
            .toLowerCase()
            .indexOf(filterText.replace(/\s/g, '').toLowerCase()) > -1
        ) {
          item.style.display = '';
        } else {
          item.style.display = 'none';
        }
      }
      return item;
    });
  };

  const handleOnKeyDown = (ev: React.KeyboardEvent<HTMLElement>) => {
    if (dropdownVisible && listRefs.current) {
      let newFocus = focus;
      if (ev.keyCode === 38) {
        newFocus--;
      } else if (ev.keyCode === 40) {
        newFocus++;
      }
      if (newFocus <= listRefs.current.length - 1) {
        const focusItem = listRefs.current[newFocus];
        focusItem && focusItem.focus();
        setFocus(newFocus);
      }
    }
  };

  const [filteredChildren] = React.useState(findListItems(children));

  return (
    <>
      <div className={styles.searchBoxWrapper}>
        <SearchBox
          {...rest}
          type={'search'}
          className={classnames(searchFieldClasses(props).main, className)}
          value={value}
          onChange={(ev, newValue) => {
            if (newValue) {
              setSearchResult(newValue);
            }
            setValue(newValue);
            onChange && onChange(ev, newValue);
          }}
          onKeyDown={ev => handleOnKeyDown(ev)}
        />
      </div>
      {dropdownVisible && (
        <div
          className={classnames(
            searchFieldClasses(props).searchListDropdown,
            styles.searchMenuDropdown
          )}
        >
          {filteredChildren}
        </div>
      )}
    </>
  );
};

export default SearchMenu;
