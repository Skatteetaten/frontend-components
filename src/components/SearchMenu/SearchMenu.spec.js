import React from 'react';
import { matches } from './../utils/test-utils';
import SearchMenu from './SearchMenu';

describe('SearchMenu komponent', () => {
  it('matcher snapshot', () => {
    matches(
      <SearchMenu>
        <ul>
          <li>skatt</li>
          <li>annen ting</li>
          <li>listeelement</li>
          <li>noe helt annet</li>
        </ul>
      </SearchMenu>
    );
  });
});
