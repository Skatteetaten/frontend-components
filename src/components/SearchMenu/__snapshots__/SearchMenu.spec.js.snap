// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchMenu komponent matcher snapshot 1`] = `
Array [
  <div
    className="searchBoxWrapper-40"
  >
    <div
      className="ms-SearchBox SkeSearchField-47"
      onFocusCapture={[Function]}
      role="search"
    >
      <div
        aria-hidden={true}
        className="ms-SearchBox-iconContainer iconContainer-48"
        onClick={[Function]}
      >
        <i
          aria-hidden={true}
          className="ms-SearchBox-icon icon-53"
          data-icon-name="Search"
        >
          Ôçâ
        </i>
      </div>
      <input
        className="ms-SearchBox-field field-51"
        id="SearchBox0"
        onChange={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        role="searchbox"
        type="search"
        value=""
      />
    </div>
  </div>,
  <div
    className="searchListDropdown-45 searchMenuDropdown-41"
  >
    <ul>
      <li
        className=""
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
      >
        skatt
      </li>
      <li
        className=""
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
      >
        annen ting
      </li>
      <li
        className=""
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
      >
        listeelement
      </li>
      <li
        className=""
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
      >
        noe helt annet
      </li>
    </ul>
  </div>,
]
`;
